### DISCUSSION TITLE
How should we handle react-router loader delays?

### DISCUSSION THREAD

**Discussion by annjawn** [remix-run/remix#5433] [Q&A]
My [route loader](https://reactrouter.com/en/main/route/loader) needs to make async calls, i am noticing that there is a delay when the user clicks a `<Link>` for a data loader route, as it waits for the fetch to complete (?). This is normally ok as long as we have a way to feedback the users through a loading icon on the new screen or on the screen where routing originated. Things get even more weird when the links are side menu, there is a natural delay and it feels like the user click didn't register. Is there a way to handle this to show a loading screen to the user while the async loader is doing it's thing. Ideally, I would like the page to be loaded with a rendered loading icon. Something like -

```
{
    path: "/",
    element: <App />,
    errorElement: <ErrorPage />,
    loader: async()  => {
      return fakeDb.from("teams").select("*");  
    } ,
    loadingElement: <Loading />,
}
```
At it's current state, what is the recommended way to handle this?

**Comment by pawelblaszczyk5** [comment]
To create loading state in the new route you can use the defer feature https://reactrouter.com/en/main/guides/deferred
And for creating an indicator of navigation in progress you can use https://reactrouter.com/en/main/hooks/use-navigation
These links are to react router docs but they're applicable in both Remix apps and apps with just react router
